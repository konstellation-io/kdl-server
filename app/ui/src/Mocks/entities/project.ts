import { memberMe, member1, member2 } from './member';
import { RepositoryType } from '../../Graphql/types/globalTypes';

const toolUrls = {
  knowledgeGalaxy: 'https://knowledgeGalaxy',
  gitea: 'https://gitea',
  filebrowser: 'https://filebrowser',
  mlflow: 'https://mlflow',
  vscode: 'https://vscode',
};

export const project1 = {
  id: 'projectId1',
  name: 'projectName1',
  description: 'projectDescription1',
  favorite: false,
  creationDate: '2020-02-02',
  lastActivationDate: '2020-02-02',
  repository: {
    type: RepositoryType.EXTERNAL,
    url: 'https://my-super-url.com',
    error: false,
  },
  needAccess: false,
  archived: false,
  error: null,
  toolUrls,
  members: [memberMe, member1, member2],
  __typename: 'Project',
};

export const project2 = {
  id: 'projectId2',
  name: 'projectName2',
  description: 'projectDescription2',
  favorite: true,
  creationDate: '2020-03-03',
  lastActivationDate: '2020-03-03',
  repository: null,
  needAccess: false,
  archived: false,
  error: null,
  toolUrls,
  members: [member1, member2],
  __typename: 'Project',
};

export const projectNoAccess = {
  id: 'projectId3',
  name: 'projectName3',
  description: 'projectDescription3',
  favorite: false,
  creationDate: '2020-02-02',
  lastActivationDate: '2020-02-02',
  repository: null,
  needAccess: true,
  archived: false,
  error: null,
  toolUrls,
  members: [member1, member2],
  __typename: 'Project',
};

export const projectArchived = {
  id: 'projectId4',
  name: 'projectName4',
  description: 'projectDescription4',
  favorite: true,
  creationDate: '2020-03-03',
  lastActivationDate: '2020-03-03',
  repository: null,
  needAccess: false,
  archived: true,
  error: null,
  toolUrls,
  members: [member1, member2],
  __typename: 'Project',
};
