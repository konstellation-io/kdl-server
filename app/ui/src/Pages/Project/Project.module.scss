@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';
@import 'Styles/shadows';

.container {
  flex-grow: 1;
  position: relative;
  width: 100%;
  height: 100%;
  background-color: palette(base, 600);

  display: flex;
  overflow: hidden;
}

.contentLayer {
  height: 100%;
  width: 100%;
  position: relative;
  overflow: auto;
}
.panelLayer {
  position: absolute;
  height: 100%;
  width: 100%;
  pointer-events: none;
  z-index: 6;
}

.panels {
  @include glow(3, black);
  display: flex;
  justify-content: flex-end;
  height: 100%;
}

.content {
  flex-grow: 1;
  height: 100%;
}
