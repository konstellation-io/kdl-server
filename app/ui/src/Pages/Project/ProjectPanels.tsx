import React, { useEffect, useState } from 'react';
import {
  memberDetails,
  primaryPanel,
  resourceDetails,
  secondaryPanel,
} from 'Graphql/client/cache';
import usePanel, { PanelType } from 'Graphql/client/hooks/usePanel';

import { GetProjects_projects } from 'Graphql/queries/types/GetProjects';
import KGResults from './panels/KGResults/KGResults';
import MemberDetails from './panels/MemberDetails/MemberDetails';
import { PANEL_ID } from 'Graphql/client/models/Panel';
import Panel from 'Components/Layout/Panel/Panel';
import ProjectSettings from './panels/ProjectSettings/ProjectSettings';
import ResourceDetails from './pages/KG/components/ResourceDetails/ResourceDetails';
import UpdateProjectDescription from './panels/UpdateProjectDescription/UpdateProjectDescription';
import UpdateRepository from './panels/UpdateRepository/UpdateRepository';
import styles from './Project.module.scss';
import useMemberDetails from 'Graphql/client/hooks/useMemberDetails';
import { useReactiveVar } from '@apollo/client';
import useResourceDetails from 'Graphql/client/hooks/useResourceDetails';

const defaultPanel = 'settings';

export interface ProjectRoute {
  openedProject: GetProjects_projects;
}
function ProjectPanels({ openedProject }: ProjectRoute) {
  const panel1Data = useReactiveVar(primaryPanel);
  const panel2Data = useReactiveVar(secondaryPanel);

  const memberDetailsData = useReactiveVar(memberDetails);
  const resourceDetailsData = useReactiveVar(resourceDetails);

  const { closePanel: panel1Close } = usePanel(PanelType.PRIMARY);
  const { closePanel: panel2Close } = usePanel(PanelType.SECONDARY);
  const { unselectMemberDetails } = useMemberDetails();
  const { unselectResourceDetails } = useResourceDetails();

  // Stores last opened tab inside project settings panel. When you reopen
  // this panel, last opened tab will remain opened.
  const [settingsOpenedTab, setSettingsOpenedTab] = useState(0);

  // Opening a level 1 panel closes previous level 2 panels
  useEffect(() => {
    if (panel1Data) panel2Close();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [panel1Data]);

  function closeMemberInfoPanel() {
    panel2Close();
    unselectMemberDetails();
  }

  function closeResourceInfoPanel() {
    panel2Close();
    unselectResourceDetails();
  }

  const panels: { [key in PANEL_ID]: JSX.Element | null } = {
    [PANEL_ID.SETTINGS]: (
      <ProjectSettings
        project={openedProject}
        settingsOpenedTab={settingsOpenedTab}
        setSettingsOpenedTab={setSettingsOpenedTab}
      />
    ),
    [PANEL_ID.REPOSITORY_INFO]: (
      <UpdateRepository project={openedProject} close={panel2Close} />
    ),
    [PANEL_ID.PROJECT_DESCRIPTION]: (
      <UpdateProjectDescription project={openedProject} close={panel2Close} />
    ),
    [PANEL_ID.MEMBER_INFO]: memberDetailsData && (
      <MemberDetails
        member={memberDetailsData}
        close={closeMemberInfoPanel}
        projectId={openedProject.id}
      />
    ),
    [PANEL_ID.KG_RESULTS]: <KGResults />,
    [PANEL_ID.KG_RESULT_DETAILS]: resourceDetailsData && (
      <ResourceDetails
        resource={resourceDetailsData}
        onClose={closeResourceInfoPanel}
      />
    ),
  };

  return (
    <div className={styles.panels}>
      <Panel
        title={panel1Data?.title}
        show={!!panel1Data}
        close={panel1Close}
        noShrink={!!panel1Data?.fixedWidth}
        theme={panel1Data?.theme}
        size={panel1Data?.size}
      >
        {panels[panel1Data?.id || defaultPanel]}
      </Panel>
      <Panel
        title={panel2Data?.title}
        show={!!panel2Data}
        close={panel2Close}
        noShrink={!!panel2Data?.fixedWidth}
        theme={panel2Data?.theme}
        size={panel2Data?.size}
      >
        {panels[panel2Data?.id || defaultPanel]}
      </Panel>
    </div>
  );
}

export default ProjectPanels;
