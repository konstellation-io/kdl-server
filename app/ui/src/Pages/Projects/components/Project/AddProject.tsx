import { Link } from 'react-router-dom';
import ROUTE from 'Constants/routes';

import IconAdd from '@material-ui/icons/Add';
import * as React from 'react';
import styles from './Project.module.scss';

const AddProject = () => (
  <div className={styles.addProjectContainer} data-testid="add-project">
    <Link to={ROUTE.NEW_PROJECT}>
      <div className={styles.addProjectBg}>
        <div className={styles.addProjectBgUpper} />
        <div className={styles.addProjectBgLower} />
      </div>
    </Link>
    <div className={styles.addProjectLabel}>
      <IconAdd className="icon-regular" />
      <p>ADD PROJECT</p>
    </div>
  </div>
);

export default AddProject;
