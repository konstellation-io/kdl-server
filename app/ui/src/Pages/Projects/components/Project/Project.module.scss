@import 'Styles/colors';
@import 'Styles/mixins';
@import 'Styles/variables';
@import 'Styles/borders';
@import 'Styles/shadows';

$label-height: 3 * $grid-unit;
$shape-height: 20 * $grid-unit;

$min-shape-width: 75 * $grid-unit;
$max-shape-width: 100 * $grid-unit;
$shape-width: 40vw;

$band-height: 5 * $grid-unit;
$big-square-base: 10 * $grid-unit;
$small-square-base: 30px;

.container {
  position: relative;
  margin: 0 0 2 * $grid-unit 5 * $grid-unit;
  flex-shrink: 0;
  pointer-events: none;

  &.archived {
    .name,
    .description {
      color: font-color(regular);
    }

    .warning {
      display: none;
    }
  }

  &:hover {
    .bg {
      box-shadow: inset 0 0 0 3px transparent,
        0 2px 3 * $grid-unit 0 palette(highlight, 900);
      background-color: palette(highlight, 900);
      border-color: palette(highlight);
    }

    .content * {
      color: palette(highlight);
    }
  }
}

.sup,
.inf {
  position: relative;
  height: $shape-height / 2;
  width: $shape-width;
  min-width: $min-shape-width;
  max-width: $max-shape-width;
}

.bg {
  background-color: palette(base, 400);
  width: 100%;
  height: 100%;
  overflow: hidden;
  box-shadow: inset 0 0 0 4px palette(base);
  border: solid 1px palette(base, 800);
  pointer-events: all;
  cursor: pointer;

  transition: background-color ease 0.4s, box-shadow ease 0.4s,
    border-color ease 0.4s;
}

.bgBand {
  position: absolute;
  left: 0;
  background-color: palette(base);
  height: $band-height / 2;
  width: 100%;
}

.content {
  position: absolute;
  top: 0;
  left: 8 * $grid-unit;
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: calc(100% - #{$band-height / 2});
  width: calc(100% - #{12 * $grid-unit});

  .contactInfo {
    @include font-small;

    pointer-events: all;
    cursor: pointer;
    display: flex;
    align-items: center;
    color: font-color(light);
    margin-left: $grid-unit;
  }
}

.sup {
  .bg {
    transform: skew(-45deg);
  }

  .bgBand {
    bottom: 0;
  }

  .content {
    left: 5 * $grid-unit;
  }
}

.inf {
  .bg {
    transform: skew(45deg);
  }

  .bgBand {
    top: 0;
  }

  .content {
    top: $band-height / 2;
    display: flex;
    flex-direction: column;
    justify-content: center;
  }
}

.name {
  @include font-subheader;
  color: font-color(light);

  margin: 0 2 * $grid-unit 0 3 * $grid-unit;

  transition: color ease 0.4s;
}

.description {
  @include font-small;
  color: font-color(regular);

  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex-shrink: 0;
  pointer-events: all;
  cursor: pointer;

  margin: 0;
  transition: color ease 0.4s;
}

.band {
  $side: $band-height / 2;

  position: absolute;
  top: ($shape-height - $band-height) / 2;
  left: ($big-square-base + $grid-unit) / 2 - 1px;
  display: flex;
  align-items: center;
  height: $band-height;
  width: calc(#{$shape-width} - #{8 * $grid-unit});
  min-width: $min-shape-width - 8 * $grid-unit;
  max-width: $max-shape-width - 8 * $grid-unit;
  padding-right: 3 * $grid-unit;

  background-color: palette(base);

  clip-path: polygon(0 100%, 0 0, 100% 0, calc(100% - #{$side}) 50%, 100% 100%);

  .otherLabels {
    display: flex;
    margin-right: auto;

    .labelArchived,
    .labelNoAccess {
      @include font-button;

      text-align: center;
      line-height: $label-height;
      height: $label-height;
      min-width: 100px;
      padding: 0 5 * $grid-unit;
      border-radius: 3 * $grid-unit;
      z-index: 1;
      margin-left: 1.5 * $grid-unit;
    }

    .labelArchived {
      color: font-color(dark);
      background-color: palette(lowlight);
    }

    .labelNoAccess {
      color: font-color(lowlight);
      background-color: palette(base, 800);
    }
  }
}

.warning {
  @include font-button;
  color: palette(alert);
  margin-right: 2 * $grid-unit;
}

.label {
  @include font-button;
  color: font-color(dark);
  text-align: center;
  line-height: $label-height;

  background-color: palette(feedback, 700);
  height: $label-height;
  min-width: 150px;
  padding: 0 5 * $grid-unit;

  border-bottom-right-radius: 3 * $grid-unit;
  border-top-right-radius: 3 * $grid-unit;

  z-index: 1;
}

.square {
  position: absolute;
  top: $big-square-base / 2;
  left: -3 * $grid-unit + 2px;
  width: $big-square-base;
  height: $big-square-base;
  z-index: 1;

  background-color: palette(base, 700);
  transform: rotate(45deg);

  .repoType {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%) rotate(-45deg);

    opacity: 0.8;
  }
}

.addProjectContainer {
  position: relative;
  margin: 0 0 2 * $grid-unit 5 * $grid-unit;
  flex-shrink: 0;

  &:hover {
    .addProjectBg > div {
      border: 1px solid palette(highlight);
    }

    .addProjectLabel {
      color: font-color(highlight);
    }
  }

  .addProjectBg {
    cursor: pointer;

    .addProjectBgUpper,
    .addProjectBgLower {
      height: $shape-height / 2;
      width: $shape-width;
      min-width: $min-shape-width;
      max-width: $max-shape-width;

      transition: border ease 0.4s, background-color ease 0.4s;
      border: 1px dashed font-color(regular);
    }

    .addProjectBgUpper {
      transform: skew(-45deg);
      border-bottom-width: 0 !important;
      border-top-color: palette(base, 400);
    }

    .addProjectBgLower {
      transform: skew(45deg);
      border-bottom-color: palette(base, 400);
      border-top-width: 0 !important;
    }
  }

  .addProjectLabel {
    @include center;

    @include font-button;
    color: font-color(regular);

    position: absolute;
    display: flex;
    align-items: center;
    pointer-events: none;

    transition: color ease 0.4s;

    p {
      margin-left: 2 * $grid-unit;
    }
  }
}

.creationDate {
  @include font-caption;
  color: font-color(lowlight);
  display: flex;
  flex-direction: row;
  width: 100%;
  margin-top: 0.5 * $grid-unit;

  p {
    margin: 0;
    padding: 0;
  }

  .creationDateLabel {
    color: font-color(regular);
    margin-right: $grid-unit;

    transition: color ease 0.2s;
  }

  .date {
    transition: color ease 0.2s;
  }
}

.disabled {
  &,
  .bg,
  .description {
    pointer-events: none !important;
  }

  .repoType {
    opacity: 0.4;
  }

  .bg {
    background-color: palette(base, 500);
  }

  .band {
    background-color: palette(base, 400);
  }

  .name {
    color: font-color(regular);
  }

  .container:hover {
    .bg {
      box-shadow: inherit;
      background-color: palette(base, 500);
      border-color: palette(base, 800);
    }

    .content * {
      color: font-color(regular);
    }
    .date {
      color: font-color(lowlight);
    }
    .contactInfo  {
      color: font-color(highlight);
    }
  }

  .addProjectContainer:hover {
    .addProjectBg > div {
      border: inherit;
    }

    .addProjectLabel {
      color: inherit;
    }
  }
}
