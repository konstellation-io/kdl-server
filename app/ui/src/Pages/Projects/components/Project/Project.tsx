import { AccessLevel } from 'Graphql/types/globalTypes';
import ROUTE, { buildRoute } from 'Constants/routes';
import React, { FC } from 'react';
import { capitalize } from 'lodash';
import RepositoryTypeComponent, {
  LOCATION,
} from 'Pages/NewProject/pages/Repository/components/RepositoryTypeComponent/RepositoryTypeComponent';

import { GetProjects_projects } from 'Graphql/queries/types/GetProjects';
import { Link } from 'react-router-dom';
import cx from 'classnames';
import { formatDate } from 'Utils/format';
import styles from './Project.module.scss';
import { ProjectAdmins } from '../../Projects';

type BaseProps = {
  project: GetProjects_projects;
};

type Props = {
  showAdmins: (projectAdmins: ProjectAdmins) => void;
  project: GetProjects_projects;
};

function Project({ project, showAdmins }: Props) {
  const disabled = project.needAccess || project.archived;
  const Component = (
    <div
      data-testid="project"
      className={cx(styles.container, {
        [styles.archived]: project.archived,
      })}
    >
      <UpperBg project={project} showAdmins={showAdmins} />
      <LowerBg project={project} />
      <Band project={project} />
      <Square project={project} />
    </div>
  );
  return disabled ? (
    <div className={styles.disabled}>{Component}</div>
  ) : (
    <Link to={buildRoute(ROUTE.PROJECT, project.id)}>{Component}</Link>
  );
}

function UpperBg({ project, showAdmins }: Props) {
  function onContactAdmins() {
    const projectAdmins = project.members.filter((u) => u.accessLevel === AccessLevel.ADMIN).map((u) => u.user.email);

    showAdmins({
      projectName: project.name,
      administrators: projectAdmins,
    });
  }

  return (
    <div className={styles.sup}>
      <div className={styles.bg}>
        <div className={styles.bgBand} />
      </div>
      <div className={styles.content}>
        <p className={styles.name} data-testid="projectName">
          {project.name}
        </p>
        {project.needAccess && (
          <div className={styles.contactInfo} onClick={onContactAdmins} data-testid="showAdminsButton">
            Need access?
          </div>
        )}
      </div>
    </div>
  );
}

const LowerBg: FC<BaseProps> = ({ project }) => (
  <div className={styles.inf}>
    <div className={styles.bg}>
      <div className={styles.bgBand} />
    </div>
    <div className={styles.content}>
      <p className={styles.description} title={project.description}>
        {project.description}
      </p>
      <div className={styles.creationDate}>
        <p className={styles.creationDateLabel}>CREATED:</p>
        <p className={styles.date}>{formatDate(new Date(project.creationDate), true)}</p>
      </div>
    </div>
  </div>
);

const Band: FC<BaseProps> = ({ project }) => (
  <div className={styles.band}>
    <div className={styles.otherLabels}>
      {project.archived && (
        <div className={styles.labelArchived} data-testid="projectArchived">
          Archived
        </div>
      )}
      {project.needAccess && <div className={styles.labelNoAccess}>No Access</div>}
    </div>
    {project.error && <div className={styles.warning}>WARNING</div>}
  </div>
);

const Square: FC<BaseProps> = ({ project }) => (
  <div className={styles.square}>
    <div className={styles.repoType}>
      <RepositoryTypeComponent
        squareLocation={LOCATION.OUT}
        customSize={38}
        shouldAnimate={false}
      />
    </div>
  </div>
);

export default Project;
