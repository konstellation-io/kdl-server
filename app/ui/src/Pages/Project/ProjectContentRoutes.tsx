import ROUTE from 'Constants/routes';
import { Redirect, Route, Switch } from 'react-router-dom';

import { GetMe } from 'Graphql/queries/types/GetMe';
import Overview from './pages/Overview/Overview';
import { ProjectRoute } from './ProjectPanels';
import ProjectToolsRoutes from './components/ProjectToolsRoutes/ProjectToolsRoutes';
import React from 'react';
import { useQuery } from '@apollo/client';

import GetMeQuery from 'Graphql/queries/getMe';

function ProjectContentRoutes({ openedProject }: ProjectRoute) {
  const { data } = useQuery<GetMe>(GetMeQuery);
  const areToolsActive = data?.me.areToolsActive;

  function redirectIfArchived() {
    return (
      openedProject.archived && (
        <Redirect from={ROUTE.PROJECT} to={ROUTE.PROJECTS} />
      )
    );
  }

  function redirectIfToolsActives() {
    return (
      !areToolsActive &&
      [ROUTE.PROJECT_TOOL_JUPYTER, ROUTE.PROJECT_TOOL_VSCODE].map((r) => (
        <Redirect key={r} from={r} to={ROUTE.PROJECT_OVERVIEW} />
      ))
    );
  }

  return (
    <Switch>
      <Redirect exact from={ROUTE.PROJECT} to={ROUTE.PROJECT_OVERVIEW} />

      {redirectIfArchived()}
      {redirectIfToolsActives()}

      <Route
        exact
        path={ROUTE.PROJECT_OVERVIEW}
        component={() => <Overview openedProject={openedProject} />}
      />
      <Route path={ROUTE.PROJECT_TOOL} component={ProjectToolsRoutes} />
    </Switch>
  );
}

export default ProjectContentRoutes;
