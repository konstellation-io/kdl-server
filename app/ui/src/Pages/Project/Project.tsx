import { ErrorMessage, SpinnerCircular } from 'kwc';
import React, { useEffect } from 'react';
import { useQuery, useReactiveVar } from '@apollo/client';

import ProjectContentRoutes from './ProjectContentRoutes';
import ProjectNavigation from './components/ProjectNavigation/ProjectNavigation';
import Tool from './pages/Tools/components/Tool/Tool';
import ProjectPanels from './ProjectPanels';
import { RouteProjectParams } from 'Constants/routes';
import styles from './Project.module.scss';
import { useParams } from 'react-router-dom';

import { currentTool, openedProject, openedTools } from 'Graphql/client/cache';

import useTools from 'Graphql/client/hooks/useTools';
import useOpenedProject from 'Graphql/client/hooks/useOpenedProject';

import { GetProjects } from 'Graphql/queries/types/GetProjects';
import GetProjectsQuery from 'Graphql/queries/getProjects';
import useRuntime from 'Graphql/client/hooks/useRuntime';
import { GetRunningRuntime } from 'Graphql/queries/types/GetRunningRuntime';
import GetRunningRuntimeQuery from 'Graphql/queries/getRunningRuntime';
import useRuntimeLoading from 'Graphql/client/hooks/useRuntimeLoading';

function Project() {
  const { projectId } = useParams<RouteProjectParams>();
  const { data, error, loading } = useQuery<GetProjects>(GetProjectsQuery);

  const project = useReactiveVar(openedProject);
  const currentToolData = useReactiveVar(currentTool);
  const openedToolsData = useReactiveVar(openedTools);

  const { updateOpenedProject } = useOpenedProject();
  const { resetTools } = useTools();
  const { updateRunningRuntime } = useRuntime();

  const { data: dataRuntimeRunning, loading: runtimeLoading } = useQuery<GetRunningRuntime>(GetRunningRuntimeQuery);
  const { setRuntimeLoading } = useRuntimeLoading();

  useEffect(() => {
    const currentProject = data?.projects.find((p) => p.id === projectId);
    if (currentProject) {
      updateOpenedProject(currentProject);
    }
    // updateOpenedProject does not change
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data, projectId]);

  useEffect(() => {
    setRuntimeLoading(runtimeLoading ? 'unknown' : '');
    if (dataRuntimeRunning?.runningRuntime) {
      updateRunningRuntime(dataRuntimeRunning.runningRuntime);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dataRuntimeRunning, runtimeLoading]);

  useEffect(
    () => () => {
      updateOpenedProject(null);
      resetTools();
    },
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  if (loading || !project) return <SpinnerCircular />;
  if (error || !data) return <ErrorMessage />;

  return (
    <div className={styles.container}>
      <ProjectNavigation />
      <div className={styles.contentLayer}>
        <ProjectContentRoutes openedProject={project} />
        {openedToolsData.map((toolName) => (
          <Tool
            key={toolName}
            name={toolName}
            src={project.toolUrls[toolName]}
            isHidden={currentToolData !== toolName}
          />
        ))}
      </div>
      <div className={styles.panelLayer}>
        <ProjectPanels openedProject={project} />
      </div>
    </div>
  );
}

export default Project;
