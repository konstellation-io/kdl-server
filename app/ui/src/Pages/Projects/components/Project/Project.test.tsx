import * as React from 'react';
import { StaticRouter } from 'react-router';
import Project from './Project';
import { projectNoAccess } from 'Mocks/entities/project';
import { mount } from 'enzyme';

const mockShowAdmins = jest.fn();

const component = mount(
  <StaticRouter>
    <Project project={projectNoAccess} showAdmins={mockShowAdmins} />
  </StaticRouter>,
);

describe('Project component', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should show admins when clicking on admins button', function () {
    component.find('[data-testid="showAdminsButton"]').simulate('click');

    expect(mockShowAdmins).toHaveBeenCalled();
  });
});
