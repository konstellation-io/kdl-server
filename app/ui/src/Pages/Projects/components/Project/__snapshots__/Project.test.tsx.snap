// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project component Component match snapshot 1`] = `
<StaticRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    staticContext={Object {}}
  >
    <Project
      project={
        Object {
          "__typename": "Project",
          "archived": false,
          "creationDate": "2020-02-02",
          "description": "projectDescription3",
          "error": null,
          "favorite": false,
          "id": "projectId3",
          "lastActivationDate": "2020-02-02",
          "members": Array [
            Object {
              "__typename": "Member",
              "accessLevel": "ADMIN",
              "addedDate": "2020-02-02",
              "user": Object {
                "accessLevel": "VIEWER",
                "creationDate": "2020-02-02",
                "email": "userEmail1",
                "id": "userId1",
                "lastActivity": "2020-02-02",
                "username": "username1",
              },
            },
            Object {
              "__typename": "Member",
              "accessLevel": "MANAGER",
              "addedDate": "2020-02-03",
              "user": Object {
                "accessLevel": "VIEWER",
                "creationDate": "2020-02-03",
                "email": "userEmail2",
                "id": "userId2",
                "lastActivity": "2020-02-03",
                "username": "username2",
              },
            },
          ],
          "name": "projectName3",
          "needAccess": true,
          "repository": null,
          "toolUrls": Object {
            "filebrowser": "https://filebrowser",
            "knowledgeGalaxy": "https://knowledgeGalaxy",
            "mlflow": "https://mlflow",
            "vscode": "https://vscode",
          },
        }
      }
      showAdmins={[MockFunction]}
    >
      <div
        className="disabled"
      >
        <div
          className="container"
          data-testid="project"
        >
          <UpperBg
            project={
              Object {
                "__typename": "Project",
                "archived": false,
                "creationDate": "2020-02-02",
                "description": "projectDescription3",
                "error": null,
                "favorite": false,
                "id": "projectId3",
                "lastActivationDate": "2020-02-02",
                "members": Array [
                  Object {
                    "__typename": "Member",
                    "accessLevel": "ADMIN",
                    "addedDate": "2020-02-02",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-02",
                      "email": "userEmail1",
                      "id": "userId1",
                      "lastActivity": "2020-02-02",
                      "username": "username1",
                    },
                  },
                  Object {
                    "__typename": "Member",
                    "accessLevel": "MANAGER",
                    "addedDate": "2020-02-03",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-03",
                      "email": "userEmail2",
                      "id": "userId2",
                      "lastActivity": "2020-02-03",
                      "username": "username2",
                    },
                  },
                ],
                "name": "projectName3",
                "needAccess": true,
                "repository": null,
                "toolUrls": Object {
                  "filebrowser": "https://filebrowser",
                  "knowledgeGalaxy": "https://knowledgeGalaxy",
                  "mlflow": "https://mlflow",
                  "vscode": "https://vscode",
                },
              }
            }
            showAdmins={[MockFunction]}
          >
            <div
              className="sup"
            >
              <div
                className="bg"
              >
                <div
                  className="bgBand"
                />
              </div>
              <div
                className="content"
              >
                <p
                  className="name"
                  data-testid="projectName"
                >
                  projectName3
                </p>
                <div
                  className="contactInfo"
                  data-testid="showAdminsButton"
                  onClick={[Function]}
                >
                  Need access?
                </div>
              </div>
            </div>
          </UpperBg>
          <LowerBg
            project={
              Object {
                "__typename": "Project",
                "archived": false,
                "creationDate": "2020-02-02",
                "description": "projectDescription3",
                "error": null,
                "favorite": false,
                "id": "projectId3",
                "lastActivationDate": "2020-02-02",
                "members": Array [
                  Object {
                    "__typename": "Member",
                    "accessLevel": "ADMIN",
                    "addedDate": "2020-02-02",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-02",
                      "email": "userEmail1",
                      "id": "userId1",
                      "lastActivity": "2020-02-02",
                      "username": "username1",
                    },
                  },
                  Object {
                    "__typename": "Member",
                    "accessLevel": "MANAGER",
                    "addedDate": "2020-02-03",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-03",
                      "email": "userEmail2",
                      "id": "userId2",
                      "lastActivity": "2020-02-03",
                      "username": "username2",
                    },
                  },
                ],
                "name": "projectName3",
                "needAccess": true,
                "repository": null,
                "toolUrls": Object {
                  "filebrowser": "https://filebrowser",
                  "knowledgeGalaxy": "https://knowledgeGalaxy",
                  "mlflow": "https://mlflow",
                  "vscode": "https://vscode",
                },
              }
            }
          >
            <div
              className="inf"
            >
              <div
                className="bg"
              >
                <div
                  className="bgBand"
                />
              </div>
              <div
                className="content"
              >
                <p
                  className="description"
                  title="projectDescription3"
                >
                  projectDescription3
                </p>
                <div
                  className="creationDate"
                >
                  <p
                    className="creationDateLabel"
                  >
                    CREATED:
                  </p>
                  <p
                    className="date"
                  >
                    Feb 2, 2020 12:00 AM
                  </p>
                </div>
              </div>
            </div>
          </LowerBg>
          <Band
            project={
              Object {
                "__typename": "Project",
                "archived": false,
                "creationDate": "2020-02-02",
                "description": "projectDescription3",
                "error": null,
                "favorite": false,
                "id": "projectId3",
                "lastActivationDate": "2020-02-02",
                "members": Array [
                  Object {
                    "__typename": "Member",
                    "accessLevel": "ADMIN",
                    "addedDate": "2020-02-02",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-02",
                      "email": "userEmail1",
                      "id": "userId1",
                      "lastActivity": "2020-02-02",
                      "username": "username1",
                    },
                  },
                  Object {
                    "__typename": "Member",
                    "accessLevel": "MANAGER",
                    "addedDate": "2020-02-03",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-03",
                      "email": "userEmail2",
                      "id": "userId2",
                      "lastActivity": "2020-02-03",
                      "username": "username2",
                    },
                  },
                ],
                "name": "projectName3",
                "needAccess": true,
                "repository": null,
                "toolUrls": Object {
                  "filebrowser": "https://filebrowser",
                  "knowledgeGalaxy": "https://knowledgeGalaxy",
                  "mlflow": "https://mlflow",
                  "vscode": "https://vscode",
                },
              }
            }
          >
            <div
              className="band"
            >
              <div
                className="label"
              />
              <div
                className="otherLabels"
              >
                <div
                  className="labelNoAccess"
                >
                  No Access
                </div>
              </div>
            </div>
          </Band>
          <Square
            project={
              Object {
                "__typename": "Project",
                "archived": false,
                "creationDate": "2020-02-02",
                "description": "projectDescription3",
                "error": null,
                "favorite": false,
                "id": "projectId3",
                "lastActivationDate": "2020-02-02",
                "members": Array [
                  Object {
                    "__typename": "Member",
                    "accessLevel": "ADMIN",
                    "addedDate": "2020-02-02",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-02",
                      "email": "userEmail1",
                      "id": "userId1",
                      "lastActivity": "2020-02-02",
                      "username": "username1",
                    },
                  },
                  Object {
                    "__typename": "Member",
                    "accessLevel": "MANAGER",
                    "addedDate": "2020-02-03",
                    "user": Object {
                      "accessLevel": "VIEWER",
                      "creationDate": "2020-02-03",
                      "email": "userEmail2",
                      "id": "userId2",
                      "lastActivity": "2020-02-03",
                      "username": "username2",
                    },
                  },
                ],
                "name": "projectName3",
                "needAccess": true,
                "repository": null,
                "toolUrls": Object {
                  "filebrowser": "https://filebrowser",
                  "knowledgeGalaxy": "https://knowledgeGalaxy",
                  "mlflow": "https://mlflow",
                  "vscode": "https://vscode",
                },
              }
            }
          >
            <div
              className="square"
            >
              <div
                className="repoType"
              >
                <Memo(RepositoryTypeComponent)
                  customSize={38}
                  shouldAnimate={false}
                  squareLocation="in"
                >
                  <div
                    className="square in medium notAnimate"
                    style={
                      Object {
                        "height": 38,
                        "width": 38,
                      }
                    }
                  >
                    <div
                      className="s4"
                    />
                    <div
                      className="s3"
                    />
                    <div
                      className="s2"
                    />
                    <div
                      className="s1"
                    />
                  </div>
                </Memo(RepositoryTypeComponent)>
              </div>
            </div>
          </Square>
        </div>
      </div>
    </Project>
  </Router>
</StaticRouter>
`;
